client/client.go

```diff
@@ -350,33 +356,39 @@ func (c *Client) Shutdown() error {
 	return c.saveState()
 }
 
-// RPC is used to forward an RPC call to a nomad server, or fail if no servers
+// RPC is used to forward an RPC call to a nomad server, or fail if no servers.
 func (c *Client) RPC(method string, args interface{}, reply interface{}) error {
 	// Invoke the RPCHandler if it exists
 	if c.config.RPCHandler != nil {
 		return c.config.RPCHandler.RPC(method, args, reply)
 	}
 
-	// Pick a server to request from
-	server := c.rpcProxy.FindServer()
-	if server == nil {
-		return fmt.Errorf("no known servers")
+	servers := c.servers.all()
+	if len(servers) == 0 {
+		return noServers
 	}
 
-	// Make the RPC request
-	if err := c.connPool.RPC(c.Region(), server.Addr, c.RPCMajorVersion(), method, args, reply); err != nil {
-		c.rpcProxy.NotifyFailedServer(server)
-		return fmt.Errorf("RPC failed to server %s: %v", server.Addr, err)
+	var mErr multierror.Error
+	for _, s := range servers {
+		// Make the RPC request
+		if err := c.connPool.RPC(c.Region(), s.addr, c.RPCMajorVersion(), method, args, reply); err != nil {
+			errmsg := fmt.Errorf("RPC failed to server %s: %v", s.addr, err)
+			mErr.Errors = append(mErr.Errors, errmsg)
+			c.logger.Printf("[DEBUG] client: %v", errmsg)
+			c.servers.failed(s)
+			continue
+		}
+		c.servers.good(s)
+		return nil
 	}
-	return nil
+
+	// Force consul discovery ASAP since we have no healthy servers
```

Comment:
Removed.

It does not always force discovery. This line described a now deleted piece of code.

Currently heartbeating is the only RPC caller that immediately causes discovery on *any* error.